feat: first impl

emirfabio opened the pull request:
Adds the core functionality for fetching OpenStreetMap data via the HOTOSM Raw Data API.

Includes:
- Asynchronous API client (`RawDataAPI`)
- Main function `get_osm_data`
- Input validation (`GeometryInput`, `RequestParams`)
- Basic CLI interface
- Utility functions (bbox conversion, saving, tile splitting)
- Initial integration tests (API & CLI)
- Example usage script

kshitijrajsharma commented on osm_data_client/api.py:
```diff
@@ -0,0 +1,230 @@
+            base_api_url: Base URL for the Raw Data API
+        """
+        self.base_api_url = base_api_url
+        self.headers = {
```
Also add access-token option in the headers and let people supply if they have it , by default API is public but with this token ability they will be able to bypass the export limit , have some staff privilege to queues based on their access token ! We can add those advance extraction methods later on , for now just letting user pass their access token would be nice if they have it ! For eg : FMTM might require priority queue

emirfabio replied:
Again, totally makes sense.

kshitijrajsharma commented on osm_data_client/api.py:
```diff
@@ -0,0 +1,230 @@
+                try:
+                    async with session.get(
+                        url=f"{self.base_api_url}{task_link}", headers=self.headers
+                    ) as response:
```
You can add logs if pooling status changes ! It will change from submitted to started when tasks starts

spwoodcock replied:
Yeah I agree - adding logs throughout the lib would be a really nice change 👍 

In each file:

```python
import logging
# other imports
log = logging.getLogger(__name__)
```

Then any upstream project using this can hook into logs.

Here you could add `log.debug("xxx")` to provide the response each time, with a `log.info("xxx")` after SUCCESS or FAILED.

For the CLI, you could instantiate a logger to the terminal stdout to hook into logs that way too:

```python

```
import sys
import logging

# set a --verbose -v flag
if args.verbose:
    logging.basicConfig(
        level="DEBUG",
        format=(
            "%(asctime)s.%(msecs)03d [%(levelname)s] "
            "%(name)s | %(funcName)s:%(lineno)d | %(message)s"
        ),
        datefmt="%y-%m-%d %H:%M:%S",
        stream=sys.stdout,
    )
```

kshitijrajsharma commented on pyproject.toml:
```diff
@@ -0,0 +1,20 @@
+version = "0.1.0"
+description = "Add your description here"
+readme = "README.md"
+requires-python = ">=3.13"
```
lets be little bit flexible with the python version ? lets support those application which uses older python , perhaps >= 3.10 would be okay ? what do you think

emirfabio replied:
yup totally, I'm simply not familiar with python environments, I've left default values on

kshitijrajsharma commented on osm_data_client/utils/file.py:
```diff
@@ -0,0 +1,127 @@
+    
+    return file_path
+
+def split_by_tiles(
```
This is lovely , I require this all the time .  I am just curious what's your motivation behind this implementation ? Are you using tiles somewhere already ?

emirfabio replied:
Honestly I saw the function in your implementation and it sounded conceptually cool so I tried understanding how it works. 
Might not be needed here, not sure.

kshitijrajsharma commented on osm_data_client/models.py:
```diff
@@ -0,0 +1,135 @@
+            raise ValidationError("Geometry must have a 'coordinates' field")
+        
+        valid_types = ["Polygon", "MultiPolygon"]
+        if geometry_dict["type"] not in valid_types:
```
Worth  adding validation to check crs of the input geometry !

kshitijrajsharma commented on osm_data_client/utils/transform.py:
```diff
@@ -0,0 +1,65 @@
+    gdf.set_crs(epsg=4326, inplace=True)
+    
+    # Reproject to the target CRS
+    gdf = gdf.to_crs(epsg=int(target_crs))
```
this is good , just make sure list of supported crs are there , this function doesn't work for the local coordinate systems or utm projected crs ! They require extra additional transformation parameters which we don't wanna go through in our first implementation, like above comment you can add validation of supported crs like 3857 and 4326 perhaps ! worth check what geopandas crs projection is capable of else we will mess up its geometry

emirfabio replied:
Yup, I got totally lost in the crs concepts trying to make the tiles implementation work. 
I'll definitely make another pass to implement your suggestions while I learn more about it!

kshitijrajsharma reviewed (commented):
Bravo ! And impressive work ! Amazing

emirfabio commented on osm_data_client/api.py:
```diff
@@ -0,0 +1,230 @@
+            except Exception as ex:
+                raise DownloadError(f"Error downloading snapshot: {str(ex)}") from ex
+    
+    def _extract_zip_data(self, zip_data: bytes, file_name: str) -> Union[Dict[str, Any], List[Dict[str, Any]], Dict[str, Union[str, bytes]]]:
```
Totally makes sense. 

Thinking about it, the basic implementation could always return the zip file bytes in a result wrapper class. 
Without adding much complexity. 
Then specific result handling based on size and output format could be re-added back in as extensions.

So that if the user _does_ want the client to handle the unpacking of data it can but it's not forced behaviour.

What do you think ?

emirfabio replied:
> For shapefiles it would be larger ! I don't like shapefiles either but unfortunately there are fortune of people who still use it and one of the popular file formats in raw-data-api

Problems like these would be handled by targeted, swappable implementations

spwoodcock replied:
### Memory Issue

- For large file unzipping, I would recommend this nice package: https://github.com/uktrade/stream-unzip.
- A very memory efficient approach to stream unzip the files from zip to disk.

### Default unzipping

- By default the API uses `bind_zip=True` for all filetypes right?
- But as this is a python lib (not direct download) and we can do neat things like stream unzip as above, we could override this behaviour.
- I think we should have some sane defaults configured for each filetype, plus an override to `force_unzip` and set `bind_zip=False`.
  - Shapefile: no (as comprises many files)
  - KML: yes
  - Mbtiles yes
  - FlatGeoBuf: yes
  - CSV: yes
  - GeoPackage: yes
  - PGDUMP: no
  - GeoJSON: yes
- So in reality I think most file types will be unzipped.
- The zip also contains a metadata JSON file if I'm not mistaken right? (off the top of my head)
- So perhaps instead of downloading that, we could simply read the metadata from the zip, and print to terminal as a DEBUG level log? (this way the user can see the info if they need, but for most they probably don't want it).

emirfabio replied:
Only thing I didn't really get: 
I can send bind_zip=False as part of the request to get a streamed response ? Is that correct?

spwoodcock replied:
Yes that's true! I neglected that part 😅 

Basically what we want is the file on the users filesystem, without the zip 👍

emirfabio commented on osm_data_client/api.py:
```diff
@@ -0,0 +1,230 @@
+                    data=json.dumps(payload),
+                    headers=self.headers,
+                ) as response:
+                    response_data = await response.json()
```
Oh for sure, totally agree!

emirfabio replied:
> You can add logs if pooling status changes ! It will change from submitted to started when tasks starts

Together with this it can give the user useful feedback on the request status

emirfabio commented:
> Bravo ! And impressive work ! Amazing

Thank you!!

spwoodcock commented on pyproject.toml:
```diff
@@ -0,0 +1,20 @@
+    "pytest-asyncio>=0.26.0",
+]
+dev = [
+    "geopandas>=1.0.1",
```
I like the approach of making these dev, the requiring the user to have them installed upstream to use them.

But note it does mean there may be version incompatibility. The user might have a version installed that doesn't work with the code written, as we have no dependency locking/solving for these deps. They **could** install using `pip install `raw-data-api-py[dev]` and this would avoid it.

I would make a note to just say, the optional functionality using geopandas was only tested on `v1.0.1`.

spwoodcock reviewed (changes requested):
Fantastic work!!

Really loving this 😄

emirfabio commented on osm_data_client/test.py:
A bit of both, it was the first testing script i wrote, committed by accident, i'll remove it hehe.

emirfabio commented on pyproject.toml:
```diff
@@ -0,0 +1,20 @@
+readme = "README.md"
+requires-python = ">=3.13"
+dependencies = [
+    "aiohttp>=3.11.16",
```
Makes sense!
ty for the explanation

emirfabio commented:
Helloooo @spwoodcock and @kshitijrajsharma! 👋

I've implemented the suggestions from the previous review. Here's what's been added:

- Memory management improvements with stream-unzip for large files
- Added proper logging throughout with appropriate levels
- Implemented queue position tracking and status changes during polling
- Added access token support in headers
- Created format-specific extraction policies (zip vs. extract) with defaults
- Special handling for shapefiles with dedicated directories
- Added proper error messages throughout

Let me know if there's anything else I should address!

Thanks for the feedback so far, really appreciate it!

spwoodcock commented:
Wow you absolutely killed it Emir - this is a very nice lib!

spwoodcock commented on osm_data_client/cli.py:
```diff
@@ -0,0 +1,234 @@
+    """
+    try:
+        # Prepare geometry
+        if args.bbox:
```
Minor pendantic note: in the geo world a bbox would typically only be a box shape. A better name for this might be --bounds or even just --geojson

emirfabio replied:
for sure, i'll change it.
